### PR TITLE
[Flink AI Flow] Add Workflow Meta to AIFlow

### PR DESCRIPTION
**Pull request by jiangxin369**
## What is the purpose of the change

Add workflow meta to AIFlow and provide related operation methods.

## Brief change log

*(for example:)*
  - Add metadata class
  - Add database model and operation methods
  - Add protos
  - Add implementation of server
  - Add implementation of client with both Python and 

## Verifying this change

This change added tests.

### CODE REVIEW

File: flink-ai-flow/ai_flow/test/store/common.py [hunk 1]
```diff
@@ -149,6 +149,60 @@ def test_update_dataset(self):
...
+
+    def test_double_register_workflow(self):
+        project_response = self.store.register_project(name='project', uri='www.code.com')
+        project_response2 = self.store.register_project(name='project2', uri='www.code.com')
+        self.store.register_workflow(name='workflow', project_id=project_response.uuid)
+        self.store.register_workflow(name='workflow', project_id=project_response2.uuid)
+        self.assertRaises(AIFlowException, self.store.register_workflow, name='workflow',
+                          project_id=project_response.uuid)
+
+    def test_list_workflow(self):
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/test/store/common.py:176]:
```suggestion
    def test_list_workflows(self):
```

File: flink-ai-flow/ai_flow/client/__init__.py [hunk 2]
```diff
@@ -212,6 +213,40 @@ def delete_project_by_name(project_name) -> Status:
...
+
+def get_workflow_by_name(project_name: Text, workflow_name: Text) -> Optional[WorkflowMeta]:
+    return get_ai_flow_client().get_workflow_by_name(project_name=project_name, workflow_name=workflow_name)
+
+
+def get_workflow_by_id(workflow_id: int) -> Optional[WorkflowMeta]:
+    return get_ai_flow_client().get_workflow_by_id(workflow_id)
+
+
+def list_workflows(project_name: Text, page_size: int, offset: int) -> Optional[List[WorkflowMeta]]:
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/client/__init__.py:228]: `list_workflows` interface should provide the function that lists workflows without paging.

File: flink-ai-flow/ai_flow/endpoint/client/metadata_client.py [hunk 3]
```diff
@@ -682,3 +683,91 @@ def delete_artifact_by_name(self, artifact_name) -> Status:
...
+        """
+        request = metadata_service_pb2.IdRequest(id=workflow_id)
+        response = self.metadata_store_stub.getWorkflowById(request)
+        return _unwrap_workflow_response(response)
+
+    def list_workflows(self, project_name: Text, page_size: int, offset: int) -> Optional[List[WorkflowMeta]]:
+        """
+        List all workflows of the specific project
+
+        :param project_name: the name of project which contains the workflow
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/endpoint/client/metadata_client.py:730]: Add the comments for the params `page_size` and `offset`.

File: flink-ai-flow/ai_flow/endpoint/client/metadata_client.py [hunk 4]
```diff
@@ -682,3 +683,91 @@ def delete_artifact_by_name(self, artifact_name) -> Status:
...
+        """
+        request = metadata_service_pb2.ListWorkflowsRequest(project_name=project_name,
+                                                            page_size=page_size,
+                                                            offset=offset)
+        response = self.metadata_store_stub.listWorkflows(request)
+        return _unwrap_workflow_list_response(response)
+
+    def delete_workflow_by_name(self, project_name: Text, workflow_name: Text) -> Status:
+        """
+        Delete the workflow by project and workflow name
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/endpoint/client/metadata_client.py:740]:
```suggestion
        Delete the workflow by specific project and workflow name
```

File: flink-ai-flow/ai_flow/endpoint/client/metadata_client.py [hunk 5]
```diff
@@ -682,3 +683,91 @@ def delete_artifact_by_name(self, artifact_name) -> Status:
...
+        return _unwrap_delete_response(response)
+
+    def update_workflow(self, workflow_name: Text, project_name: Text,
+                        properties: Properties = None) -> Optional[WorkflowMeta]:
+        """
+        Update the workflow
+
+        :param workflow_name: the workflow name
+        :param project_name: the name of project which contains the workflow
+        :param properties: (Optional) the properties needs to be updated
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/endpoint/client/metadata_client.py:767]:
```suggestion
        :param properties: (Optional) the properties need to be updated
```

File: flink-ai-flow/ai_flow/java/client/src/main/java/com/aiflow/client/MetadataClient.java [hunk 6]
```diff
@@ -971,4 +973,111 @@ public Status deleteArtifactByName(String artifactName) throws Exception {
...
+
+    /***
+     * Get a workflow by specific project name and workflow name
+     *
+     * @param projectName   The name of project which contains the workflow
+     * @param workflowName  The workflow name
+     * @return WorkflowMeta object registered in Metadata Store
+     */
+    public WorkflowMeta getWorkflowByName(String projectName, String workflowName) throws Exception {
+        WorkflowNameRequest request = WorkflowNameRequest.newBuilder().setProjectName(projectName).setWorkflowName(workflowName ).build();
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/java/client/src/main/java/com/aiflow/client/MetadataClient.java:1003]:
```suggestion
        WorkflowNameRequest request = WorkflowNameRequest.newBuilder().setProjectName(projectName).setWorkflowName(workflowName).build();
```

File: flink-ai-flow/ai_flow/java/client/src/main/java/com/aiflow/client/MetadataClient.java [hunk 7]
```diff
@@ -971,4 +973,111 @@ public Status deleteArtifactByName(String artifactName) throws Exception {
...
+    public List<WorkflowMeta> listWorkflows(String projectName, Long pageSize, Long offset) throws Exception {
+        ListWorkflowsRequest request = ListWorkflowsRequest.newBuilder()
+                .setProjectName(projectName).setPageSize(pageSize).setOffset(offset).build();
+        Response response = metadataServiceStub.listWorkflows(request);
+        WorkflowListProto.Builder builder = WorkflowListProto.newBuilder();
+        return StringUtils.isEmpty(metadataDetailResponse(response, this.parser, builder)) ? null : buildWorkflowMetas(builder.build());
+    }
+
+    /***
+     * Delete the workflow by project and workflow name
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/java/client/src/main/java/com/aiflow/client/MetadataClient.java:1041]:
```suggestion
     * Delete the workflow by specific project and workflow name
```

File: flink-ai-flow/ai_flow/meta/workflow_meta.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+                 properties: Properties = None,
+                 create_time: int = None,
+                 update_time: int = None,
+                 uuid: int = None
+                 ) -> None:
+        """
+
+        :param name: the workflow name
+        :param project_id: the uuid of project which contains this workflow
+        :param create_time: create time represented as milliseconds since epoch.
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/meta/workflow_meta.py:39]: Add the comment for the param `properties`.

File: flink-ai-flow/ai_flow/metadata_store/service/service.py [hunk 9]
```diff
@@ -407,3 +408,44 @@ def deleteArtifactById(self, request, context):
     def deleteArtifactByName(self, request, context):
         status = self.store.delete_artifact_by_name(request.name)
         return _wrap_delete_response(status)
+
+    @catch_exception
+    def registerWorkflow(self, request, context):
+        workflow = transform_workflow_meta(request.workflow)
+        response = self.store.register_workflow(name=workflow.name,
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/metadata_store/service/service.py:415]: Does this need to handle the field `properties` like `properties = None if request.properties == {} else request.properties`?

File: flink-ai-flow/ai_flow/scheduler/scheduling_service.py [hunk 10]
```diff
@@ -75,6 +75,7 @@ def submitWorkflow(self, request, context):
                     result=ResultProto(status=StatusProto.ERROR,
                                        error_message='{}, {} do not exist!'.format(project_name,
                                                                                    workflow.workflow_name)))
+
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/scheduler/scheduling_service.py:78]: No need to change this file.

File: flink-ai-flow/ai_flow/store/abstract_store.py [hunk 11]
```diff
@@ -295,6 +295,72 @@ def delete_project_by_name(self, project_name):
...
+        """
+        List all workflows of the specific project
+
+        :param project_name: the name of project which contains the workflow
+        """
+        pass
+
+    def delete_workflow_by_name(self, project_name, workflow_name):
+        """
+        Delete the workflow by project and workflow name
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/store/abstract_store.py:339]:
```suggestion
        Delete the workflow by specific project and workflow name
```

File: flink-ai-flow/ai_flow/store/mongo_store.py [hunk 12]
```diff
@@ -478,6 +479,136 @@ def delete_project_by_name(self, project_name) -> Status:
...
+        if len(workflow_result) == 0:
+            return None
+        workflows = []
+        for workflow in workflow_result:
+            workflows.append(ResultToMeta.result_to_workflow_meta(workflow))
+        return workflows
+
+    def delete_workflow_by_name(self, project_name, workflow_name) -> Status:
+        """
+        Delete the workflow by project and workflow name
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/store/mongo_store.py:550]:
```suggestion
        Delete the workflow by specific project and workflow name
```

File: flink-ai-flow/ai_flow/store/sqlalchemy_store.py [hunk 13]
```diff
@@ -749,6 +750,150 @@ def delete_model_version_relation_by_version(self, version, model_id) -> Status:
...
+        """
+        Delete the workflow by specific id
+
+        :param workflow_id: the uuid of workflow
+        """
+        workflow = self.get_workflow_by_id(workflow_id)
+        project = self.get_project_by_id(workflow.project_id)
+        if workflow is None or project is None:
+            return Status.ERROR
+        return self.delete_workflow_by_name(project_name=project.name, workflow_name=workflow.name)
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/store/sqlalchemy_store.py:868]: Why not delete workflow directly by uuid? IMO, UUID is primary key for the `workflow` table.

File: flink-ai-flow/ai_flow/test/endpoint/test_client.py [hunk 14]
```diff
@@ -216,6 +216,63 @@ def test_update_project(self):
...
+
+    def test_double_register_workflow(self):
+        project_response = client.register_project(name='project', uri='www.code.com')
+        project_response2 = client.register_project(name='project2', uri='www.code.com')
+        client.register_workflow(name='workflow', project_id=project_response.uuid)
+        client.register_workflow(name='workflow', project_id=project_response2.uuid)
+        self.assertRaises(AIFlowException, client.register_workflow, name='workflow',
+                          project_id=project_response.uuid)
+
+    def test_list_workflow(self):
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow/test/endpoint/test_client.py:244]:
```suggestion
    def test_list_workflows(self):
```

**Review by SteNicholas — CHANGES_REQUESTED** [review]

File: flink-ai-flow/ai_flow/metadata_store/service/service.py (hunk 9 above)
Comment (jiangxin369) [reply flink-ai-flow/ai_flow/metadata_store/service/service.py:415]: Yes, it's handled in upper `transform_workflow_meta`

**Review by SteNicholas — APPROVED** [review]

**Review by wuchaochen — APPROVED** [review]

**Pull request merged**